### PR TITLE
Detect if the GridEngine worker thread has crashed to prevent hanging the workflow

### LINKED ISSUE #4856: Failed `sbatch` may hang a workflow
Mira noticed that a workflow she was running reported a failed `sbatch` command, and then reported a long run of `squeue` commands succeeding while apparently doing nothing.

It may be that a failed `sbatch` can kill a job submission thread that we need to be alive for the workflow not to get stuck.

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1533)
┆Issue Number: TOIL-1533

### PR DESCRIPTION
**Pull request by stxue1**
Resolves #4856 

A failed `sbatch` command would cause the background thread (used to be called worker) to crash and die, hanging the entire workflow as there was no logic to detect that the thread is now gone, and python's `Threads` doesn't carry through exceptions.

The `squeue` commands being reported is due to the command call function spamming the debug log: https://github.com/DataBiosphere/toil/blob/78fe93db1f10fb7ba77db90694549fecc0f6470d/src/toil/lib/misc.py#L141

This is useful from a debugging perspective, but can be vague and spammy at times, such as in this case, where the `squeue` commands were due to the leader checking for deadlocks (and overlong jobs): https://github.com/DataBiosphere/toil/blob/78fe93db1f10fb7ba77db90694549fecc0f6470d/src/toil/leader.py#L814

I'm hesitant to add more debug messages as the leader main loop will call these functions repeatedly. Ideally there is some way to have Toil log something even lower priority than debug which will then be used for maintenance. It is already at debug level though, so it shouldn't be too user facing.
## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * Add logic to detect if the worker thread of the GridEngineBatchSystem dies unexpectedly
 * Rename GridEngineBatchSystem's Worker to GridEngineThread

## Reviewer Checklist

 * [ ] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/{cliOptions,cwl,wdl}.rst` 
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

File: src/toil/batchSystems/abstractGridEngineBatchSystem.py [hunk 1]
```diff
@@ -428,7 +434,12 @@ def killBatchJobs(self, jobIDs):
         for jobID in jobIDs:
             self.killQueue.put(jobID)
         while jobIDs:
-            killedJobId = self.killedJobsQueue.get()
+            try:
+                killedJobId = self.killedJobsQueue.get(timeout=10)
+            except queue.Empty:
```
Comment (adamnovak) [inline src/toil/batchSystems/abstractGridEngineBatchSystem.py:439]: It looks like we already had `from queue import Empty, Queue`, so we should probably use the shorter name:
```suggestion
            except Empty:
```

File: src/toil/batchSystems/abstractGridEngineBatchSystem.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 import logging
+import queue
```
Comment (adamnovak) [inline src/toil/batchSystems/abstractGridEngineBatchSystem.py:15]: If we already import `Queue` and `Empty` later, we don't need the whole module probably.
```suggestion
```

File: src/toil/batchSystems/abstractGridEngineBatchSystem.py [hunk 3]
```diff
@@ -277,14 +280,17 @@ def run(self):
                 while self._runStep():
                     pass
             except Exception as ex:
-                logger.error("GridEngine like batch system failure", exc_info=ex)
-                raise
+                self.exception = ex
+                logger.error("GridEngine like batch system failure: %s", ex)
+                # dont raise exception as is_alive will still be set to false,
```
Comment (adamnovak) [inline src/toil/batchSystems/abstractGridEngineBatchSystem.py:285]:
```suggestion
                # don't raise exception as is_alive will still be set to false,
```

File: src/toil/lib/retry.py [hunk 4]
```diff
@@ -572,6 +572,7 @@ def old_retry(
     >>> i
     1
     """
+    timeout = timeout if timeout else DEFAULT_TIMEOUT
```
Comment (adamnovak) [inline src/toil/lib/retry.py:575]: Won't this break `timeout=0`? Maybe this should be:
```suggestion
    if timeout is None:
        timeout = DEFAULT_TIMEOUT
```

Or maybe we really want `timeout=None` to represent not using a timeout?

**Review by adamnovak — CHANGES_REQUESTED** [review]
This looks pretty good, but I'm not sure what's going on with the change to the retry timeout logic.

I'm also not 100% convinced that `GridEngineThread` is the best name for the thread that talks to the scheduler. But I don't have a better idea for a name, so maybe that will have to stay.

File: src/toil/lib/retry.py (hunk 4 above)
Comment (stxue1) [reply src/toil/lib/retry.py:575]: I think setting it to a default might be better. With the current retry code the only way to simply set an infinite timeout is to just set the timeout to a really high number.

**Review by adamnovak — APPROVED** [review]

**Pull request merged**